File: __tests__/core/mockNetwork.test.ts
import { MockNetwork, MockPaymentDeclined } from '@/core/mock/mockNetwork';

describe('MockNetwork', () => {
  it('countAndMaybeFail throws on every nth call', () => {
    const net = new MockNetwork({ minDelayMs: 0, maxDelayMs: 1, failEvery: 3 });
    net.countAndMaybeFail();
    net.countAndMaybeFail();
    expect(() => net.countAndMaybeFail()).toThrow(MockPaymentDeclined);
    net.countAndMaybeFail();
    net.countAndMaybeFail();
    expect(() => net.countAndMaybeFail()).toThrow(MockPaymentDeclined);
  });

  it('seeded delays are deterministic across instances', () => {
    const a = new MockNetwork({ minDelayMs: 1, maxDelayMs: 50 });
    const b = new MockNetwork({ minDelayMs: 1, maxDelayMs: 50 });
    expect(a.nextDelayMs()).toBe(b.nextDelayMs());
    expect(a.nextDelayMs()).toBe(b.nextDelayMs());
  });

  it('delay resolves within bounds', async () => {
    const net = new MockNetwork({ minDelayMs: 1, maxDelayMs: 2 });
    await expect(net.delay()).resolves.toBeUndefined();
  });
});

File: test/core/mock_network_test.dart
import 'package:billpayments/core/mock/mock_network.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('delay completes within configured bounds', () async {
    final network = MockNetwork(minDelayMs: 1, maxDelayMs: 2);
    final stopwatch = Stopwatch()..start();
    await network.delay();
    expect(stopwatch.elapsedMilliseconds, greaterThanOrEqualTo(1));
    expect(stopwatch.elapsedMilliseconds, lessThan(2000));
  });

  test('countAndMaybeFail throws on every nth call', () {
    final network = MockNetwork(minDelayMs: 0, maxDelayMs: 1, failEvery: 3);
    network.countAndMaybeFail();
    network.countAndMaybeFail();
    expect(network.countAndMaybeFail, throwsA(isA<MockPaymentDeclined>()));
    network.countAndMaybeFail();
    network.countAndMaybeFail();
    expect(network.countAndMaybeFail, throwsA(isA<MockPaymentDeclined>()));
  });

  test('seeded delays are deterministic across instances', () async {
    final a = MockNetwork(minDelayMs: 1, maxDelayMs: 50);
    final b = MockNetwork(minDelayMs: 1, maxDelayMs: 50);
    expect(a.nextDelayMs(), b.nextDelayMs());
    expect(a.nextDelayMs(), b.nextDelayMs());
  });
}
